checker/src/main/java/org/checkerframework/checker/determinism/DeterminismQualifierPolymorphism.java

```diff
@@ -0,0 +1,145 @@
+        super(env, factory);
+        this.factory = factory;
+    }
+
+    /**
+     * Replaces {@code @PolyDet} in {@code type} with the instantiations in {@code replacements}.
+     * Replaces {@code @PolyDet("up")} with {@code @NonDet} if it resolves to {@code OrderNonDet}.
+     * Replaces {@code @PolyDet("down")} with {@code @Det} if it resolves to {@code OrderNonDet}.
+     * Replaces {@code @PolyDet("use")} with the same annotation that {@code @PolyDet} resolves to.
+     *
+     * @param type annotated type whose poly annotations are replaced
+     * @param replacements mapping from polymorphic annotation to instantiation
+     */
+    @Override
+    protected void replace(
+            AnnotatedTypeMirror type, AnnotationMirrorMap<AnnotationMirrorSet> replacements) {
+        boolean isPolyUp = false;
+        boolean isPolyDown = false;
+        if (type.hasAnnotation(factory.POLYDET_UP)) {
+            isPolyUp = true;
+            type.replaceAnnotation(factory.POLYDET);
+        } else if (type.hasAnnotation(factory.POLYDET_DOWN)) {
+            isPolyDown = true;
+            type.replaceAnnotation(factory.POLYDET);
+        }
+
+        if (type.hasAnnotation(factory.POLYDET) || type.hasAnnotation(factory.POLYDET_USE)) {
+            AnnotationMirrorSet quals = replacements.get(factory.POLYDET);
+            type.replaceAnnotations(quals);
+        } else {
+            for (Map.Entry<AnnotationMirror, AnnotationMirrorSet> pqentry :
+                    replacements.entrySet()) {
+                AnnotationMirror poly = pqentry.getKey();
+                if (type.hasAnnotation(poly)) {
+                    type.removeAnnotation(poly);
+                    AnnotationMirrorSet quals = pqentry.getValue();
+                    type.replaceAnnotations(quals);
+                }
+            }
+        }
+
+        if (isPolyUp
+                && (type.hasAnnotation(factory.ORDERNONDET)
+                        || type.hasAnnotation(factory.NONDET))) {
+            replaceForPolyUpOrDown(type, factory.NONDET);
+        }
+        if (isPolyDown
+                && (type.hasAnnotation(factory.ORDERNONDET) || type.hasAnnotation(factory.DET))) {
+            replaceForPolyUpOrDown(type, factory.DET);
+        }
+    }
+
+    /**
+     * This method replaces the annotation of {@code type} and all the nested type arguments of
+     * {@code type} with {@code replaceType}.
+     *
+     * @param type the polymorphic type to be replaced
+     * @param replaceType the type to be replaced with
+     */
+    private void replaceForPolyUpOrDown(AnnotatedTypeMirror type, AnnotationMirror replaceType) {
+        type.replaceAnnotation(replaceType);
+        if (!(factory.isCollection(TypesUtils.getTypeElement(type.getUnderlyingType()).asType())
+                || factory.isIterator(
+                        TypesUtils.getTypeElement(type.getUnderlyingType()).asType()))) {
```

Comment:
This code appears multiple times.  Why are you going from a TypeMirror to a TypeElement back to a TypeMirror?